libs/stokesian_dynamics/sd.hpp

```diff
@@ -0,0 +1,1478 @@
+    // (that is, the distance between particle centers)
+    DEVICE_FUNC void operator()(std::size_t part_id) {
+        // Fill the self mobility terms 
+        // These are the sub-tensors of the grand mobility matrix as shown
+        // in equation (A 1) that are located on the diagonal.
+        // This functor fills in the sub-tensors for one single particle and
+        // has to be executed for each particle.
+
+        // mob_a, mob_c and mob_m are templates for those sub-tensors that just
+        // have to be rescaled by a constant factor.
+        static constexpr multi_array<T, 3, 3> const mob_a = {
+            // clang-format off
+            1, 0, 0,
+            0, 1, 0,
+            0, 0, 1
+            // clang-format on
+        };
+        static constexpr multi_array<T, 3, 3> const mob_c = {
+            // clang-format off
+            3./4., 0    , 0    ,
+            0    , 3./4., 0    ,
+            0    , 0    , 3./4.
+            // clang-format on
+        };
+        static constexpr multi_array<T, 5, 5> const mob_m = {
+            // clang-format off
+            9./5. , 0    , 0    , 0    , 9./10.,
+            0     , 9./5., 0    , 0    , 0     ,
+            0     , 0    , 9./5., 0    , 0     ,
+            0     , 0    , 0    , 9./5., 0     ,
+            9./10., 0    , 0    , 0    , 9./5.
+            // clang-format on
+        };
+
+
+        // Compute where the self mobility submatrices of the current particle
+        // are located in the grand mobility matrix.
+        // For velocities/forces, there are 6 independent components.
+        // (3 translation and 3 rotation)
+        // For shear rate/stresslets there are 5 independent components.
+        std::size_t const ph1 = 6 * part_id;
+        std::size_t const ph2 = ph1 + 3;
+        std::size_t const ph3 = 5 * part_id;
+
+        // These are the non-dimensionalizations as stated in the paragraph
+        // below equation (A 1).
+        auto const visc1 = T{M_1_PI / 6. / eta / a(part_id)}; 
+        auto const visc2 = T{visc1 / a(part_id)};
+        auto const visc3 = T{visc2 / a(part_id)};
+
+        // Now put the entries into the grand mobility matrix.
+        for (std::size_t i = 0; i < 3; ++i) {
+            zmuf(ph1 + i, ph1 + i) = visc1 * mob_a(i, i);
+            zmuf(ph2 + i, ph2 + i) = visc3 * mob_c(i, i);
+        }
+
+        for (std::size_t i = 0; i < 5; ++i) {
+            for (std::size_t j = 0; j < 5; ++j) {
+                zmes(ph3 + i, ph3 + j) = visc3 * mob_m(i, j);
+            }
+        }
+    }
+};
+
+template <typename Policy, typename T>
+struct mobility<Policy, T, false> {
+    device_matrix_view<T, Policy> zmuf, zmus, zmes;
+    device_matrix_view<T, Policy> const pd;
+    device_matrix_view<std::size_t, Policy> const part_id;
+    device_vector_view<T, Policy> const a;
+    T const eta;
+    int const flg;
+
+    // Determine the pair contribution
+    DEVICE_FUNC void operator()(std::size_t pair_id) {
```

Comment:
I know this function is somewhat bulky. I just thought about how to split it up, but it is rather difficult. I would separate it into three steps:

1. prepare a bunch of numbers needed for step 2
2. compute all necessary sub-tensors
3. distribute all those values from step 2 into the correct locations of the grand mobility matrix.

Steps 1 and 2 are closely connected and to would be awkward to split them up. Step 3 could be separated from steps 1 and 2, because the only thing step 3 needs is a set of sub-tensors generated by step 2.